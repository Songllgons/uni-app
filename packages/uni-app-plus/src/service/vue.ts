export { initApp } from '@dcloudio/uni-vue'
