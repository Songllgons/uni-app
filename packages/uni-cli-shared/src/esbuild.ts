import type { BuildOptions } from 'esbuild'
import path from 'path'

export function transformWithEsbuild(
  code: string,
  filename: string,
  options: BuildOptions
) {
  options.stdin = {
    contents: code,
    resolveDir: path.dirname(filename),
  }
  return import('esbuild').then((esbuild) => {
    return esbuild.build(options)
  })
}

export function esbuild(options: BuildOptions) {
  return import('esbuild').then((esbuild) => {
    return esbuild.build(options)
  })
}
