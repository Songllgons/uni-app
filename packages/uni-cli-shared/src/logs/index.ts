export { formatErrMsg, formatInfoMsg, formatWarnMsg } from './format'

type LogType = 'error' | 'warn' | 'info' | 'log'
let lastType: LogType | undefined
let lastMsg: string | undefined

export function resetOutput(type: LogType) {
  if (type === lastType) {
    lastType = undefined
    lastMsg = ''
  }
}

export function output(type: LogType, msg: string) {
  if (type === lastType && msg === lastMsg) {
    return
  }
  lastMsg = msg
  lastType = type
  const method = type === 'info' ? 'log' : type
  console[method](msg)
}
