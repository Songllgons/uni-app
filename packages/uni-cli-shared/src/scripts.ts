import { hasOwn } from '@vue/shared'
import fs from 'fs'

interface Package {
  'uni-app'?: {
    scripts?: {
      [name: string]: {
        title?: string
        BROWSER?: 'Chrome' | 'Firefox' | 'IE' | 'Edge' | 'Safari' | 'HBuilderX'
        env?: {
          UNI_PLATFORM: UniApp.PLATFORM
          [name: string]: string
        }
        define: {
          [name: string]: boolean
        }
      }
    }
  }
}

export function parseScripts(name: string, pkgPath: string) {
  if (!fs.existsSync(pkgPath)) {
    return
  }
  const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8')) as Package

  const scripts = pkg['uni-app']?.scripts || {}
  const options = scripts[name]
  if (!options) {
    return
  }
  if (!options.env?.UNI_PLATFORM) {
    console.error(
      `package.json->uni-app->scripts->${name}->env->UNI_PLATFORM is required`
    )
    process.exit(0)
  }
  const { UNI_PLATFORM, ...define } = options.env
  const context = options.define || {}
  // 补充当前编译环境未定义的其他编译环境 define
  Object.keys(scripts).forEach((scriptName) => {
    if (scriptName !== name) {
      const scriptDefine = scripts[scriptName].define || {}
      Object.keys(scriptDefine).forEach((key) => {
        if (!hasOwn(context, key)) {
          context[key] = false
        }
      })
    }
  })
  return {
    name: name,
    platform: UNI_PLATFORM,
    define,
    context,
  }
}
