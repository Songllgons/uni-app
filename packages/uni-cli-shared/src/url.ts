import base64url from 'base64url'

export function encodeBase64Url(str: string) {
  return base64url.encode(str)
}

export function decodeBase64Url(str: string) {
  return base64url.decode(str)
}
