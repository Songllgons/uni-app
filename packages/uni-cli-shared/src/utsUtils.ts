import { matchEasycom } from './easycom'
import { camelize, capitalize, normalizeNodeModules, removeExt } from './utils'

export function matchUTSComponent(tag: string) {
  const source = matchEasycom(tag)
  return !!(source && source.includes('uts-proxy'))
}

export function genUTSClassName(fileName: string, prefix: string = 'Gen') {
  return (
    prefix +
    capitalize(
      camelize(
        verifySymbol(
          removeExt(
            normalizeNodeModules(fileName)
              .replace(/[\/|_]/g, '-')
              .replace(/-+/g, '-')
          )
        )
      )
    )
  )
}

function isValidStart(c: string): boolean {
  return !!c.match(/^[A-Za-z_-]$/)
}

function isValidContinue(c: string): boolean {
  return !!c.match(/^[A-Za-z0-9_-]$/)
}

function verifySymbol(s: string) {
  const chars = Array.from(s)

  if (isValidStart(chars[0]) && chars.slice(1).every(isValidContinue)) {
    return s
  }

  const buf: string[] = []
  let hasStart = false

  for (const c of chars) {
    if (!hasStart && isValidStart(c)) {
      hasStart = true
      buf.push(c)
    } else if (isValidContinue(c)) {
      buf.push(c)
    }
  }

  if (buf.length === 0) {
    buf.push('_')
  }

  return buf.join('')
}
