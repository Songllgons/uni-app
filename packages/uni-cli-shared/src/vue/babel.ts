import type * as BabelCore from '@babel/core'
import type { PluginObj } from '@babel/core'
import { isBuiltInComponent } from '@dcloudio/uni-shared'

export function transformUniH5Jsx({ types }: typeof BabelCore): PluginObj {
  return {
    name: 'babel-plugin-uni-h5-jsx',
    visitor: {
      JSXOpeningElement({ node: { name } }) {
        if (types.isJSXIdentifier(name) && isBuiltInComponent(name.name)) {
          name.name = 'v-uni-' + name.name
        }
      },
    },
  }
}
