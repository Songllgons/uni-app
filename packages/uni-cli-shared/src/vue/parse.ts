import {
  type AttributeNode,
  type ElementNode,
  NodeTypes,
  type ParentNode,
  type RootNode,
  type TemplateChildNode,
} from '@vue/compiler-core'
import MagicString from 'magic-string'
import { isElementNode, parseVue } from '../vite/utils/ast'

const BLOCK_RE = /<\/block>/
const WXS_LANG_RE = /lang=["|'](renderjs|wxs|sjs)["|']/
const WXS_ATTRS = ['wxs', 'renderjs', 'sjs']

export function parseVueCode(code: string, isNVue = false) {
  const hasBlock = BLOCK_RE.test(code)
  const hasWxs = WXS_LANG_RE.test(code)
  if (!hasBlock && !hasWxs) {
    return { code }
  }
  const errors: SyntaxError[] = []
  const files: string[] = []
  let ast = parseVue(code, errors)
  if (hasBlock) {
    code = parseBlockCode(ast, code)
    // 重新解析新的 code
    ast = parseVue(code, errors)
  }
  if (!isNVue && hasWxs) {
    const wxsNodes = parseWxsNodes(ast)
    code = parseWxsCode(wxsNodes, code)
    // add watch
    for (const wxsNode of wxsNodes) {
      const srcProp = wxsNode.props.find(
        (prop) => prop.type === NodeTypes.ATTRIBUTE && prop.name === 'src'
      ) as AttributeNode | undefined
      if (srcProp && srcProp.value) {
        files.push(srcProp.value.content)
      }
    }
  }
  return { code, files, errors }
}

function traverseChildren({ children }: ParentNode, blockNodes: ElementNode[]) {
  children.forEach((node) => traverseNode(node, blockNodes))
}

function traverseNode(
  node: RootNode | TemplateChildNode,
  blockNodes: ElementNode[]
) {
  if (isElementNode(node) && node.tag === 'block') {
    blockNodes.push(node)
  }
  if (
    node.type === NodeTypes.IF_BRANCH ||
    node.type === NodeTypes.FOR ||
    node.type === NodeTypes.ELEMENT ||
    node.type === NodeTypes.ROOT
  ) {
    traverseChildren(node, blockNodes)
  }
}

export function parseBlockCode(ast: RootNode, code: string) {
  const blockNodes: ElementNode[] = []
  traverseNode(ast, blockNodes)
  if (blockNodes.length) {
    return parseBlockNode(code, blockNodes)
  }
  return code
}
const BLOCK_END_LEN = '</block>'.length
const BLOCK_START_LEN = '<block'.length

function parseBlockNode(code: string, blocks: ElementNode[]) {
  const magicString = new MagicString(code)
  blocks.forEach(({ loc }) => {
    const startOffset = loc.start.offset
    const endOffset = loc.end.offset
    magicString.overwrite(
      startOffset,
      startOffset + BLOCK_START_LEN,
      '<template'
    )
    magicString.overwrite(endOffset - BLOCK_END_LEN, endOffset, '</template>')
  })
  return magicString.toString()
}

export function parseWxsNodes(ast: RootNode) {
  return ast.children.filter(
    (node) =>
      node.type === NodeTypes.ELEMENT &&
      node.tag === 'script' &&
      node.props.find(
        (prop) =>
          prop.name === 'lang' &&
          prop.type === NodeTypes.ATTRIBUTE &&
          prop.value &&
          WXS_ATTRS.includes(prop.value.content)
      )
  ) as ElementNode[]
}

export function parseWxsCode(wxsNodes: ElementNode[], code: string) {
  if (wxsNodes.length) {
    code = parseWxsNode(code, wxsNodes)
  }
  return code
}

const SCRIPT_END_LEN = '</script>'.length
const SCRIPT_START_LEN = '<script'.length

function parseWxsNode(code: string, nodes: ElementNode[]) {
  const magicString = new MagicString(code)
  nodes.forEach(({ loc, props }) => {
    const langAttr = props.find((prop) => prop.name === 'lang') as AttributeNode
    const moduleAttr = props.find(
      (prop) => prop.name === 'module'
    ) as AttributeNode
    const startOffset = loc.start.offset
    const endOffset = loc.end.offset
    const lang = langAttr.value!.content
    const langStartOffset = langAttr.loc.start.offset
    magicString.overwrite(
      startOffset,
      startOffset + SCRIPT_START_LEN,
      '<' + lang
    ) // <renderjs or <wxs
    magicString.overwrite(
      langStartOffset,
      langStartOffset + ('lang="' + lang + '"').length,
      ''
    ) // remove lang="renderjs" or lang="wxs"
    magicString.overwrite(
      endOffset - SCRIPT_END_LEN,
      endOffset,
      '</' + lang + '>'
    ) //</renderjs> or </wxs>

    if (moduleAttr) {
      const moduleStartOffset = moduleAttr.loc.start.offset
      magicString.overwrite(
        moduleStartOffset,
        moduleStartOffset + 'module'.length,
        'name'
      ) // module="echarts" => name="echarts"
    }
  })
  return magicString.toString()
}
