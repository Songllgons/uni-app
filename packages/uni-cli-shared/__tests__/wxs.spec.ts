import { parseVue } from '../src/vite'
import { parseWxsCode, parseWxsNodes } from '../src/vue'

describe('wxs', () => {
  test('parseWxsCode', () => {
    const renderjsCode = `<template><view></view><view></view></template>
  <script>
  export default {}
  </script>
  <script lang="renderjs" module="echarts">
  export default{
      mounted(){
          console.log('mounted')
      }
  }
  </script>
  <style>
  .content {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
  }
  </style>
  `
    expect(
      parseWxsCode(parseWxsNodes(parseVue(renderjsCode, [])), renderjsCode)
    ).toMatchSnapshot()
    const wxsCode = `<template><view></view><view></view></template>
    <script>
    export default {}
    </script>
    <script lang="wxs" module="echarts">
    export default{
        mounted(){
            console.log('mounted')
        }
    }
    </script>
    <style>
    .content {
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
    }
    </style>
    `
    expect(
      parseWxsCode(parseWxsNodes(parseVue(wxsCode, [])), wxsCode)
    ).toMatchSnapshot()
  })
})
