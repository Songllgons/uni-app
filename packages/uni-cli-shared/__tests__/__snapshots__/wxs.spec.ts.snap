// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wxs parseWxsCode 1`] = `
"<template><view></view><view></view></template>
  <script>
  export default {}
  </script>
  <renderjs  name="echarts">
  export default{
      mounted(){
          console.log('mounted')
      }
  }
  </renderjs>
  <style>
  .content {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
  }
  </style>
  "
`;

exports[`wxs parseWxsCode 2`] = `
"<template><view></view><view></view></template>
    <script>
    export default {}
    </script>
    <wxs  name="echarts">
    export default{
        mounted(){
            console.log('mounted')
        }
    }
    </wxs>
    <style>
    .content {
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: center;
    }
    </style>
    "
`;
